src/browser/pages/vscode.ts

```diff
@@ -160,9 +156,72 @@ export function setBodyBackgroundToThemeBackgroundColor(document: Document, loca
   return null
 }
 
-try {
+export function registerLoadBundleOnNlsConfig(nlsConfig: NlsConfiguration, base: string) {
+  const bundles: {
+    [key: string]: string
+  } = {}
+
+  type LoadBundleCallback = (_: undefined, result?: string) => void
+
+  nlsConfig.loadBundle = (bundle: string, _language: string, cb: LoadBundleCallback): void => {
+    const result = bundles[bundle]
+    if (result) {
+      return cb(undefined, result)
+    }
+    // FIXME: Only works if path separators are /.
+    const path = nlsConfig._resolvedLanguagePackCoreLocation + "/" + bundle.replace(/\//g, "!") + ".nls.json"
+    fetch(`${base}/vscode/resource/?path=${encodeURIComponent(path)}`)
+      .then((response) => response.json())
+      .then((json) => {
+        bundles[bundle] = json
+        cb(undefined, json)
+      })
+      .catch(cb)
+  }
+}
+
+/**
+ * A helper function to encapsulate all the
+ * logic used in this file.
+ *
+ * It does the following:
+ * - grabs the options using getOptions
+ * - gets the nlsConfig
+ * - registers loadBundle on nlsConfig
+ * - gets the loader
+ * - sets loader on self.require
+ * - sets the body background color to match the theme
+ *
+ * We purposely include all of this in a single function
+ * so that it's easier to test.
+ */
+export function main() {
+  const options = getOptions()
+  const nlsConfig = getNlsConfiguration(document)
+
+  if (nlsConfig._resolvedLanguagePackCoreLocation) {
+    registerLoadBundleOnNlsConfig(nlsConfig, options.base)
+  }
+
+  const loader = getLoader({
+    nlsConfig,
+    options,
+    origin: window.location.origin,
+  })
+
+  // TODO@jsjoeio
+  // I'm not sure how to properly type cast this
+  // This might be our best bet
+  // Source: https://stackoverflow.com/a/30740935
+  // This object on self.require is what configures the loader
+  // and tells it how to load files that get imported.
+  ;(self.require as FixMeLater) = loader
```

Comment:
Maybe we can copy down the `.d.ts` for the loader to get the right type.

Reply:
This might be a lot of work though, not sure. We also might need to load `loader.js` first and then do `require.config(config)` instead.

Reply:
> Maybe we can copy down the .d.ts

I'll take a look.

> This might be a lot of work though, not sure. We also might need to load loader.js first and then do require.config(config) instead.

Hmm! I'll look and if it is, we can do as a followup item

Reply:
> This might be a lot of work though, not sure. We also might need to load loader.js first and then do require.config(config) instead.

To be honest, I don't fully understand how this all works, so I'm going to punt and say this should be a follow-up item.

Reply:
Though after re-reading your comment, I think we might be talking about two separate things. I'm trying to get the right types for `self.require` (which may not be solvable beyond using `any` and it sounds like you're talking about getting the right types for `loader`. Is that right?

https://github.com/cdr/code-server/issues/3780

Reply:
My bad, I meant to imply we can tackle this later and not in this PR but I did not make that clear!

I think they do something like this, broadly speaking:

```
const options = self.require
self.require = new requireFuncFactory(options) // This returns an IRequireFunc which is the loader
```

I thought they declared `require` as `IRequireFunc` but they do not so you can ignore my previous comment. I think we can just do something like:

```
;(self.require as unknown as IConfigurationOptions) = { options }
```

Reply:
![image](https://user-images.githubusercontent.com/3806031/126227356-199e6a9d-74df-46c5-b346-bb1e793c5881.png)

I'm having trouble finding `IConfigurationOptions` but I did something similar

Reply:
Oops nevermind ignore me on the second part, you already linked to that `.d.ts` file in a comment so I think you were just saying it's not in VS Code and not that you couldn't find `IConfigurationOptions` at all.